### DISCUSSION TITLE
@riverpod generator generates the wrong type of provider

### DISCUSSION THREAD

**Discussion by ptheofan** [rrousselGit/riverpod#2352] [Q&A] [answered]
I have provider that holds all the logged in user information, preferences, etc. This provider I don't even want to die. Should exist for the lifetime of the app.

The declaration looks like this
```dart
@riverpod
class AppUser extends _$AppUser {
  @override
  AppUserDto build() => const AppUserDto();

  /// setters/getters and other functions
}
```

The generator creates the following for this provider
```dart
/// See also [AppUser].
@ProviderFor(AppUser)
final appUserProvider =
    AutoDisposeNotifierProvider<AppUser, AppUserDto>.internal(
  AppUser.new,
  name: r'appUserProvider',
  debugGetCreateSourceHash:
      const bool.fromEnvironment('dart.vm.product') ? null : _$appUserHash,
  dependencies: null,
  allTransitiveDependencies: null,
);
```

As it turns out, the moment user goes from login page to another page the provider gets disposed and recreated when the other page is open.

What am I doing wrong?

info: when adding ref.keepAlive() in the build method the provider is no longer disposed and works fine. But I would love to know if I am doing something wrong as I would have expected the generator to create a provider without the `autoDispose` modifier.

**Comment by AhmedLSayed9** [comment, marked as answer]
`@riverpod` by default generates an autoDispose provider.
If you want to generate a non-autoDispose provider you can use `@Riverpod(keepAlive: true)`

**Reply by ptheofan** [reply]
omg, oh yes! `@Riverpod` ftw!
Might I recommend an update of the documentation, the docs-v2 have the `@riverpod` there.

**Reply by AhmedLSayed9** [reply]
Yes because that's the default and recommended to be used.
`@Riverpod(keepAlive: true)` should be added to the `.autoDispose` modifier section or something. Docs are currently WIP and you can also contribute if you want.
